### ISSUE TITLE
image optimizer returns 400 when path contains unicode characters

### ISSUE THREAD

**Issue by stefanprobst** [vercel/next.js#27210]
### What version of Next.js are you using?

11.0.1

### What version of Node.js are you using?

14.17.2

### What browser are you using?

firefox

### What operating system are you using?

ubuntu

### How are you deploying your application?

vercel

### Describe the Bug

in a production build, the image optimizer returns with status code 400, and error message "The requested resource isn't a valid image" when the image src contains non-ascii unicode characters.

### Expected Behavior

image optimizer should handle paths with all allowed unicode characters

### To Reproduce

1. clone https://github.com/stefanprobst/issue-next-image-optimizer and `yarn install`
2. run `yarn build && yarn start`
3. there should be an image visible at <http://localhost:3000>, but there isn't. the network panel reports status code 400, error message "The requested resource isn't a valid image." which comes from [here](https://github.com/vercel/next.js/blob/7644c17f43b35a5b2b56ee99906fe89f7350ffba/packages/next/server/image-optimizer.ts#L309)
4. note that when running `yarn dev` the image is visible

**Comment by stefanprobst** [comment]
also happening with 11.0.2-canary.16

**Comment by styfle** [comment]
I wasn't able to reproduce on macOS (tried with Firefox and Chrome using Node 14). I see the image with 200 status.

Perhaps there is something different about Ubuntu filesystem causing the issue.

![image](https://user-images.githubusercontent.com/229881/127352354-0cdf6020-0941-4663-a275-7ee9c8b17987.png)

**Comment by stefanprobst** [comment]
thanks for looking into this.

i can still reliably reproduce this, both on a local linux install (w/ firefox+chrome), and when deployed to our k8s - but not when deployed to vercel.

![Screenshot_20210728_181406](https://user-images.githubusercontent.com/20753323/127358678-204ddc3d-5bd7-46ee-8fbb-64f45a4d5580.png)

![Screenshot_20210728_181824](https://user-images.githubusercontent.com/20753323/127359204-71155f09-7f15-4e17-abb5-42ef98c86958.png)

**Comment by styfle** [comment]
"The requested resource isn't a valid image" is the error when the upstream image has an invalid `Content-Type` header

https://github.com/vercel/next.js/blob/36b81f989c01bf8bbf823de8d2677d98b82390fe/packages/next/server/image-optimizer.ts#L324-L328

What's the `Content-Type` when you visit `http://localhost:3000/leärn.webp`?

Is the image even visible when visiting it directly like that?

**Comment by stefanprobst** [comment]
yes when visiting directly the image is visible and has `image/webp` content type:
![Screenshot_20210729_064046](https://user-images.githubusercontent.com/20753323/127432768-b942dc1d-8a8a-4981-99bd-8d87137a3440.png)

**Comment by stefanprobst** [comment]
https://github.com/vercel/next.js/blob/36b81f989c01bf8bbf823de8d2677d98b82390fe/packages/next/server/image-optimizer.ts#L287-L288

here, the detected content type is `null`, which then falls back to `mockRes.getHeader('Content-Type')` which is `text/html; charset=utf-8`

**Comment by stefanprobst** [comment]
i have now tried in a Windows 10 VM, and can reproduce the same error there as well, so it's not a linux-specific thing.

**Comment by styfle** [comment]
I verified the image bytes match the expected `[0x52, 0x49, 0x46, 0x46, ?, ?, ?, ?, 0x57, 0x45, 0x42, 0x50]` pattern 

![image](https://user-images.githubusercontent.com/229881/127516785-b7d546a5-66b2-4fe6-b1b3-af9a770cf687.png)

Furthermore, the fallback `Content-Type` header should have been assigned to the same value, `image/webp`, since you confirmed that visiting the image directly has the correct `Content-Type`.

How about if you `console.log(upstreamBuffer)` or even try `console.log(upstreamBuffer.toString('utf8'))`?

It sounds like the mock response is not routed properly and returns a HTML page instead of the image 🤔

**Comment by stefanprobst** [comment]
jep indeed: `console.log(upstreamBuffer.toString('utf8'))` prints:

<details>
<summary>a HTML page</summary>

```
<!DOCTYPE html><html><head><meta name="viewport" content="width=device-width"/><meta charSet="utf-8"/><title>404: This page could not be found</title><meta name="next-head-count" content="3"/><noscript data-n-css=""></noscript><script defer="" nomodule="" src="/_next/static/chunks/polyfills-a54b4f32bdc1ef890ddd.js"></script><script src="/_next/static/chunks/webpack-98f4799ea432c775d6c4.js" defer=""></script><script src="/_next/static/chunks/framework-56721e7fe9e004cd9e49.js" defer=""></script><script src="/_next/static/chunks/main-5cf1eb10e66e91cc19cd.js" defer=""></script><script src="/_next/static/chunks/pages/_app-3593cae11b23df12898b.js" defer=""></script><script src="/_next/static/chunks/pages/_error-28c7d26a0b1f8acb56d8.js" defer=""></script><script src="/_next/static/lBc3afPWHYSEeapBMlP9p/_buildManifest.js" defer=""></script><script src="/_next/static/lBc3afPWHYSEeapBMlP9p/_ssgManifest.js" defer=""></script></head><body><div id="__next"><div style="color:#000;background:#fff;font-family:-apple-system, BlinkMacSystemFont, Roboto, &quot;Segoe UI&quot;, &quot;Fira Sans&quot;, Avenir, &quot;Helvetica Neue&quot;, &quot;Lucida Grande&quot;, sans-serif;height:100vh;text-align:center;display:flex;flex-direction:column;align-items:center;justify-content:center"><div><style>body { margin: 0 }</style><h1 style="display:inline-block;border-right:1px solid rgba(0, 0, 0,.3);margin:0;margin-right:20px;padding:10px 23px 10px 0;font-size:24px;font-weight:500;vertical-align:top">404</h1><div style="display:inline-block;text-align:left;line-height:49px;height:49px;vertical-align:middle"><h2 style="font-size:14px;font-weight:normal;line-height:inherit;margin:0;padding:0">This page could not be found<!-- -->.</h2></div></div></div></div><script id="__NEXT_DATA__" type="application/json">{"props":{"pageProps":{"statusCode":404}},"page":"/_error","query":{},"buildId":"lBc3afPWHYSEeapBMlP9p","nextExport":true,"isFallback":false,"gip":true,"scriptLoader":[]}</script></body></html>
```

</details>

I also put a `console.trace()` here:
https://github.com/vercel/next.js/blob/c63bfbe773edd6328b0d99953334f3245bbae74d/packages/next/server/image-optimizer.ts#L259
to see where the Content-Type header gets set (not sure if that's useful):

<details>
<summary>Trace: Setting Content-Type header</summary>

```
    at Writable.mockRes.setHeader (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/image-optimizer.js:261:33)
    at Object.sendPayload (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/send-payload.js:42:13)
    at Server.pipe (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/next-server.js:847:38)
    at async Object.fn (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/next-server.js:645:21)
    at async Router.execute (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/router.js:205:32)
    at async Server.run (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/next-server.js:811:29)
    at async Server.handleRequest (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/next-server.js:281:20)
    at async Object.imageOptimizer (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/image-optimizer.js:280:17)
    at async Router.execute (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/router.js:205:32)
    at async Server.run (/home/stefan/Development/playground/issue-next-image-optimizer/node_modules/next/dist/server/next-server.js:811:29)
```

</details>

**Comment by styfle** [comment]
That HTML is the 404 page so definitely something wrong with routing here.

I tested on Windows with both Chrome and Firefox but I'm unable to reproduce.

Perhaps it has something to do with the language/locale of your browser?

I have "English (United States)" aka `en-US` selected. What do you have selected?

### Windows with Chrome
![windows-chrome](https://user-images.githubusercontent.com/229881/127568745-e500e3c1-4060-49d7-878d-d8de1a2f40b9.png)

### Windows with Firefox
![windows-firefox](https://user-images.githubusercontent.com/229881/127568973-fa73847a-1219-4013-bda9-9d1d73784a40.png)

**Comment by stefanprobst** [comment]
> Perhaps it has something to do with the language/locale of your browser?

i have `de`, and my system locale is `de_AT`. however, checking with firefox nightly (without any languages installed), started with `LANG=C firefox-trunk` gives the same error:

![Screenshot_20210730_080808](https://user-images.githubusercontent.com/20753323/127608957-8130f06c-fb1f-4f1c-9a02-2886fe8af4b7.png)

**Comment by stefanprobst** [comment]
i'm now wondering why it works with `next dev`, but not with `next build && next start` - how does image optimization differ between dev and prod?

**Comment by stefanprobst** [comment]
> i have de, and my system locale is de_AT.

also tried in a docker container, which should have `en` system locale - same error

(pushed the `Dockerfile` to the reproduction repro for reference, build with `docker build -t issue-next-image-optimizer .` and run with `docker run --rm -p 3000:3000 issue-next-image-optimizer`)

**Comment by stefanprobst** [comment]
ok, so what appears to be happening is that the "public folder catchall" route does not correctly match here:
https://github.com/vercel/next.js/blob/34d418c12d593c4180f805fc51229259caa5e28a/packages/next/server/next-server.ts#L1142

this is what i get here:
![next-server-L1137](https://user-images.githubusercontent.com/20753323/127732346-977aab94-88c2-440d-9a70-5ea96c8c0184.png)

you can see that the `publicFiles` `Set` is URI-encoded, but the `/leärn.png` path is not (probably because it is taken from a query param?).

**Comment by stefanprobst** [comment]
when i `encodeURI` the `url` here:
https://github.com/vercel/next.js/blob/34d418c12d593c4180f805fc51229259caa5e28a/packages/next/server/image-optimizer.ts#L72
the repro example is working correctly for me now

**Comment by styfle** [comment]
Nice find! Would you like to submit a PR?

**Comment by stefanprobst** [comment]
sure, done in #27671

**Comment by ctjlewis** [comment]
> I wasn't able to reproduce on macOS. [...] Perhaps there is something different about Ubuntu filesystem causing the issue.

Lol, if I had a nickel.

Really good work here @stefanprobst.

**Comment by styfle** [comment]
The fix is available in [v11.0.2-canary.25](https://github.com/vercel/next.js/releases/tag/v11.0.2-canary.25)

You can try it out with `yarn add next@canary`, thanks!

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
